text/0000-placement-by-return.md

```diff
@@ -0,0 +1,986 @@
+
+## Returning unsized Results
+[returning-unsized-results]: #Returning-unsized-Results
+
+This RFC's biggest drawback is its inability to handle faillible emplacement.
+
+As mentionned above, the syntax `Box::new_with(|| foo()?)` wouldn't be applicable, even disregarding layout problems. Most likely, special-case methods would have to be added (eg `<doTheThing>_with_result`), so that the final syntax would look like:
+
+```rust
+let my_box : Box<[i32]> = Box::new_with_result(|| foo())?;
+```
+
+However, there are several obstacles to this syntax:
+
+- It would require establishing a representation for DSTs wrapped in enums. Currently the only way to make a custom DST is to add another DST at the end of a struct.
+
+  There is no way to do the same with an enum; if we wanted to implement that feature, we would need to establish semantics for how these enums interact with the rest of the language. These semantics are beyond the scope of this RFC.
+
+- A `new_with_result` is only possible if the function can internally allocate enough memory to store a result, while only keeping the range of memory storing the payload at the end of the call, discarding the discriminant and the error (and accounting for cases where the err variant of the Result is much heavier than the ok variant).
+
+  While this is feasible with Box or Rc (which could, for instance, always allocate extraneous bytes before their stored data to accomodate the discriminant), it's harder for methods such as `Vec::push_with`, which require working on contiguous memory.
+
+There are several potential solutions to the second problem:
+
+### Split the discriminant from the payload
+
+We could decide that enums (or at least a specific subset of enums, including Result, Option, and other specialized types) should be stored differently.
+
+Instead of storying the discriminant next to the payload, it would always be stored separately, eg in a register. References to these enums would be fat pointers, with the first word pointing to the payload, and the second word storing the discriminant.
+
+Functions returning `Result` could then store their payload in the return slot, while returning the discriminant directly through a register; they could even take two return slots: one for the `Ok` variant and another for the `Err` variant.
+
+Of course, a naive implementation of this change would be far-reaching, and break backwards compatibility. Among other things, it would severely impact how Results and Options and similar types are stored in containers.
+
+One possible mitigation might be treat enums as normal, contiguous chunks of data when storing them and passing them to function; and to treat them as fat pointers when returning them from functions. However, the semantics in that case are non-trivial; among other things, they might clash with future attempts to add NRVO.
```

Comment:
Potential clashes with NRVO and `write_return_with` seem more relevant in this section. Maybe even just move the "Vs alternative ABI for returning Result" section here instead- return ABI changes for enums are less an alternative to this RFC and more an extension, since they do not really address placement on their own without GCE.